Add support for NSPrivateQueueConcurrencyType

sbonami opened the pull request:
Currently, any context that uses the `NSPrivateQueueConcurrencyType` concurrency type deadlocks on `existingObjectWithID:error:` in [`insertOrUpdateObjects:ofEntity:withContext:error:completionBlock:`](https://github.com/sbonami/PFIncrementalStore/blob/master/PFIncrementalStore/PFIncrementalStore.m#L813). This is bad!

Alydyn commented:
Is this for ALL contexts that use private concurrency, or just contexts with a parent context?

In other words if a context does not have a `parentContext` and does have `NSPrivateQueueConcurrencyType`, does it still hang?

sbonami commented:
PFIS creates child `NSPrivateQueueConcurrencyType` contexts that rely on parent contexts, therefore all contexts that use the private concurrency are affected.

Commit:
Remove blocking calls causing issue with Private contexts

This commit removes thread-blocking synchronous calls to
`performBlockAndWait` that were affecting contexts with a private
concurrency type. The thread-blocking caused a cyclical deadlock when a
private context called it’s parent context.

Commit:
Extract child NSManagedObject creation into helper method

Commit:
Update spec tests for changed threading

This commit updates the spec tests due to the changes in threading as a
result of the private concurrency type bug fix.

Commit:
Add context save after sync so managed context notification kicks off

To allow notifications to be received for the parent context upon save,
this commit adds a save block.
